index.js

```diff
@@ -96,54 +135,146 @@ original error: ${e}`);
+  filteredPodcasts = Object.values(temp);
+  // console.log('done dups');
+
+  if (filepatterns.length > 0) {
+    console.log(`Exporting ${filteredPodcasts.length} of ${podcasts.length}`);
+  }
+
+  function joinPath(parts) {
+    return parts.filter((s) => s).join('/');
+  }
+
+  // Make all necessary directories.  Each podcast is in its own
+  // subdir.
   const outputDir = getOutputDirPath();
-  await fs.mkdir(outputDir, { recursive: true });
+  const allDirs = filteredPodcasts.map(p => {
+    return joinPath([outputDir, p.podcastName]);
+  });
+  const uniqueDirs = Array.from(new Set(allDirs));
+  // console.log(`Making ${uniqueDirs.length} directories ...`);
+  uniqueDirs.forEach(d => mkdirSync(d, { recursive: true }));
+  // console.log(`Done making ${uniqueDirs.length} directories.`);
+
+  let skipped = 0;
+
+  // Actual file export.
   await Promise.all(
-    filesWithDBData.map(async (podcast) => {
-      // Create an export subdir
-      let exportDirPath = outputDir;
-      const exportDir = podcast.dbMeta?.zpodcast;
-      if (exportDir) {
-        exportDirPath = `${outputDir}/${exportDir}`;
-        // Needs to be sync else the same dir can be created multiple times
-        if (!existsSync(exportDirPath)) {
-          mkdirSync(exportDirPath);
-        }
+    filteredPodcasts.map(async (p) => {
+      const parts = [outputDir, p.podcastName, p.exportFileName];
+      const newPath = joinPath(parts);
+      const logName = joinPath([p.podcastName, p.exportFileName]);
+      if (!existsSync(newPath)) {
+        console.log(`${p.fileName} -> ${logName}`);
+        await exportSingle(p, newPath);
       }
-      const date = podcast.dbMeta?.date;
-      const exportFileName = podcast.dbMeta?.zcleanedtitle // 1. from apple podcast database
-        ?? (await getMP3MetaTitle(podcast.path)) // 2. from mp3 meta data
-        ?? podcast.uuid; // 3. fallback to unreadable uuid
-      const sanitizedExportFileName = sanitize(exportFileName.substr(0, fileNameMaxLength));
-      const newPath = `${exportDirPath}/${sanitizedExportFileName}.mp3`;
-      await fs.copyFile(podcast.path, newPath);
-      console.log(`${podcast.path} -> ${newPath}`);
-      if (date) {
-        const d = new Date(date);
-        await fs.utimes(newPath, d, d);
+      else {
+        skipped += 1;  // Might not work w/ promises, but not concerned.
+        console.log(`Already have ${logName}, skipping`);
       }
     })
   );
-  console.log(`\n\nSuccessful Export to '${outputDir}' folder!`);
+
+  console.log(`\n\nExported ${filteredPodcasts.length} podcasts to '${outputDir}'`);
+  if (skipped > 0) {
+    console.log(`(skipped ${skipped}, already present)`);
+  }
   exec(`open ${outputDir}`);
 }
 
-async function main() {
+async function main(filepatterns = []) {
   const dbPodcastData = await tryGetDBPodcastsData();
-  await exportPodcasts(dbPodcastData);
+  await exportPodcasts(dbPodcastData, filepatterns);
+}
+
+// Default: return all files.
+let patterns = []
```

Comment:
nitpicking: I would move that logic to `main` to have on the top level only function call

Reply:
Yes, that's better, thanks.